crate.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/jackc/pgx"
+	"github.com/jackc/pgx/pgtype"
+	"github.com/prometheus/common/model"
+)
+
+type crateRow struct {
+	labels     model.Metric
+	labelsHash string
+	timestamp  time.Time
+	value      float64
+	valueRaw   int64
+}
+
+type crateWriteRequest struct {
+	rows []*crateRow
+}
+
+type crateReadRequest struct {
+	stmt string
+}
+
+type crateReadResponse struct {
+	rows []*crateRow
+}
+
+type crateEndpoint struct {
+	pool     *pgx.ConnPool
+	poolConf pgx.ConnPoolConfig
+}
+
+func newCrateEndpoint(ep *endpointConfig) *crateEndpoint {
+	connConf := pgx.ConnConfig{
+		Host:     ep.Host,
+		Port:     ep.Port,
+		User:     ep.User,
+		Password: ep.Password,
+		Database: ep.Database,
+	}
+	if ep.EnableTLS {
+		connConf.TLSConfig = &tls.Config{
+			ServerName:         ep.Host,
+			InsecureSkipVerify: ep.AllowInsecureTLS,
+		}
+	}
+	poolConf := pgx.ConnPoolConfig{
+		ConnConfig:     connConf,
+		MaxConnections: ep.MaxConnections,
+	}
+	return &crateEndpoint{poolConf: poolConf}
+}
+
+func (c *crateEndpoint) endpoint() endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+		// We initialize the connection pool lazily here instead of in newCrateEndpoint() so
+		// that the adapter does not crash on startup if an endpoint is unavailable.
+		if c.pool == nil {
+			pool, err := pgx.NewConnPool(c.poolConf)
+			if err != nil {
+				return nil, fmt.Errorf("error opening connection to CrateDB: %v", err)
+			}
+			c.pool = pool
+		}
+
+		switch r := request.(type) {
+		case *crateWriteRequest:
+			return nil, c.write(ctx, r)
+		case *crateReadRequest:
+			return c.read(ctx, r)
+		default:
+			panic("unknown request type")
+		}
+	}
+}
+
+func (c crateEndpoint) write(ctx context.Context, r *crateWriteRequest) error {
+	_, err := c.pool.PrepareEx(ctx, "write_statement", `INSERT INTO metrics ("labels", "labels_hash", "timestamp", "value", "valueRaw") VALUES ($1, $2, $3, $4, $5)`, nil)
```

Comment:
I think we should make the prepared statement name unique, e.g. suffix with timestamp.

Reply:
Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

Reply:
> I think we should make the prepared statement name unique, e.g. suffix with timestamp.

Reading https://godoc.org/github.com/jackc/pgx#ConnPool.PrepareEx and skimming the underlying implementation, it at least seems like it should be ok to always use the same name for the same statement string:

"PrepareEx is idempotent; i.e. it is safe to call PrepareEx multiple times with the same name and sql arguments. This allows a code path to PrepareEx and Query/Exec/Prepare without concern for if the statement has already been prepared."

> Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

I'd be surprised if you could measure any difference (haven't tried yet), especially since this only happens once per remote-write-request, not per sample. But I think it's a good idea to put it into a `const` anyway.

Reply:
Ok, cool. Sounds good to me.